internal/host/plugin/repository_host_catalog.go

```diff
@@ -143,6 +146,279 @@ func (r *Repository) CreateCatalog(ctx context.Context, c *HostCatalog, _ ...Opt
+	// updating.
+	currentCatalog, _, err := r.LookupCatalog(ctx, c.PublicId)
+	if err != nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("error looking up catalog with id %q", c.PublicId)))
+	}
+
+	if currentCatalog == nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.RecordNotFound, op, fmt.Sprintf("catalog with id %q not found", c.PublicId))
+	}
+
+	// Assert the version of the current catalog to make sure we're
+	// updating the correct one.
+	if currentCatalog.GetVersion() != version {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.VersionMismatch, op, fmt.Sprintf("catalog version mismatch, want=%d, got=%d", currentCatalog.GetVersion(), version))
+	}
+
+	// Clone the catalog so that we can set fields.
+	newCatalog := currentCatalog.clone()
+	var updateAttributes bool
+	var dbMask, nullFields []string
+	for _, f := range fieldMask {
+		switch {
+		case strings.EqualFold("name", f) && c.Name == "":
+			nullFields = append(nullFields, "name")
+			newCatalog.Name = c.Name
+		case strings.EqualFold("name", f) && c.Name != "":
+			dbMask = append(dbMask, "name")
+			newCatalog.Name = c.Name
+		case strings.EqualFold("description", f) && c.Description == "":
+			nullFields = append(nullFields, "description")
+			newCatalog.Description = c.Description
+		case strings.EqualFold("description", f) && c.Description != "":
+			dbMask = append(dbMask, "description")
+			newCatalog.Description = c.Description
+		case strings.EqualFold("attributes", strings.Split(f, ".")[0]):
+			// Flag attributes for updating. While multiple masks may be
+			// sent, we only need to do this once.
+			updateAttributes = true
+		case strings.EqualFold("secrets", f):
+			// While in a similar format, secrets are passed along
+			// wholesale (for the time being). Don't append this mask
+			// field, as secrets do not have a database entry. Clear the
+			// secrets out of the working set after.
+			newCatalog.Secrets = c.Secrets
+			c.Secrets = nil
+
+		default:
+			return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidFieldMask, op, fmt.Sprintf("invalid field mask: %s", f))
+		}
+	}
+
+	if updateAttributes && c.Attributes != nil {
+		dbMask = append(dbMask, "attributes")
+		newCatalog.Attributes, err = patchstruct.PatchBytes(newCatalog.Attributes, c.Attributes)
+		if err != nil {
+			return nil, db.NoRowsAffected, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg("error in catalog attribute JSON"))
+		}
+
+		// Patch the working set with the new attributes.
+		c.Attributes = newCatalog.Attributes
+	}
+
+	// Get the plugin client.
+	plgClient, ok := r.plugins[currentCatalog.GetPluginId()]
+	if !ok || plgClient == nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("plugin %q not available", currentCatalog.GetPluginId()))
```

Comment:
I don't think this is an invalid parameter anymore when we already have a catalog with this plugin id.  IMO it makes sense on create since we don't necessarily know if the plugin id exists, but now we know that it does.

This seems like a plugin specific error code might be good to create or make it an internal error.

Reply:
Yeah, maybe internal for now (unless we already have a class/code range for plugin errors). Will update.

Reply:
Done!